packages/http-server/src/http-server.ts

```diff
@@ -4,22 +4,24 @@
 // License text available at https://opensource.org/licenses/MIT
 
 import {createServer, Server, ServerRequest, ServerResponse} from 'http';
+import {createServer as createHttpsServer, Server as HttpsServer} from 'https';
+import * as https from 'https';
 import {AddressInfo} from 'net';
+
 import * as pEvent from 'p-event';
 
-export type HttpRequestListener = (
-  req: ServerRequest,
-  res: ServerResponse,
-) => void;
+export type RequestListener = (req: ServerRequest, res: ServerResponse) => void;
 
-/**
- * Object for specifyig the HTTP / HTTPS server options
- */
-export type HttpServerOptions = {
-  port?: number;
+export interface ListenerOptions {
   host?: string;
+  port?: number;
+}
+
+export interface HttpOptions extends ListenerOptions, https.ServerOptions {
```

Comment:
I would like to see two options interfaces, one for HTTP and another for HTTPS config. That way TypeScript can offer better auto-completion in IDE - when configuring `http` server, no HTTPS-related options should be offered for auto-completion. See https://github.com/strongloop/loopback-next/pull/1488#issuecomment-401286048 for a code snippet illustrating what I mean.

What is your reasoning for the currently proposed design with a single interface, @hacksparrow?

Reply:
There is an interface in `@loopback/rest/src/rest.server.ts` which looks like this:
```
export interface RestServerConfig
  extends RestServerOptions,
    HttpServerOptions {}
```
It requires `HttpServerOptions` to be an interface. You had suggested:

```
interface ListenerOptions {
  host?: string;
  port?: number;
}

interface HttpOptions extends ListenerOptions {
  protocol: 'http' | undefined;
}

interface HttpsOptions extends ListenerOptions, https.ServerOptions {
  protocol: 'https';
}

type HttpServerOptions = HttpOptions | HttpsOptions;
````
I like the suggestion too, but in this case `HttpServerOptions` is a type, which can't be extended by `RestServerConfig`.

@bajtos any suggestions on resolving this conflicting requirements?

Reply:
IMO, `RestServerConfig` does not have to be an interface. Have you tried the following?

```ts
export type RestServerConfig = RestServerOptions & HttpServerOptions;
```

To me, it reads pretty naturally - a rest server configuration is combination of rest-specific options with HTTP(s) server options.